hvp.php

```diff
@@ -248,21 +392,71 @@ public function saveLibraryDependencies($libraryId, $dependencies, $dependency_t
   }
 
   /**
-   * Implements saveContentData
+   * Implements updateContent
+   *
+   * Updates existing H5P content.
+   *
+   * @param array $content
+   *   An associative array containing:
+   *   - id: The content id
+   *   - params: The content in json format
+   *   - library: An associative array containing:
+   *     - libraryId: The id of the main library for this content
+   * @param int $contentMainId
+   *   Main id for the content if this is a system that supports versioning
    */
-  public function saveContentData($contentId, $contentJson, $mainJsonData, $mainLibraryId, $contentMainId = NULL) {
+  public function updateContent($content, $contentMainId = NULL) {
     global $DB;
-    
-    $DB->insert_record_raw('hvp_contents', array(
-      'id' => $contentId,
-      'content' => $contentJson,
-      'embed_type' => (isset($mainJsonData['embedTypes']) ? implode(', ', $mainJsonData['embedTypes']) : ''),
-      'library_id' => $mainLibraryId
-    ), false, false, true);
+
+    // There's always an id because the H5P entry is first saved to the
+    // 'hvp' table, and the same id is being used in the 'hvp_contents'
+    // table. Therefore we must check manually whether the row exists.
+    $existing = $DB->get_record('hvp_contents', array('id' => $content['id']));
```

Comment:
The failure when uploading a new H5P package was caused by Moodle refusing to run the `insert_record()` function because the data already had an id (which should not exists for new rows). Now the code check manually whether the row exists, and decides whether to user `insert_record()` or `update_record()` based in it.

I wonder whether the `hvp_contents` table should have a separate column (e.g `hvp_id`) that would match the `id` of the `hvp` table. And the `id` column of `hvp_contents` would be auto-incremented for each row.